torch/_subclasses/fake_tensor.py

```diff
@@ -0,0 +1,140 @@
+import torch
+
+from torch._subclasses import BaseTensor
+from torch.utils._pytree import tree_map
+from functools import partial
+from torch.fx.operator_schemas import normalize_function
+from typing import Union
+
+aten = torch.ops.aten
+
+_device_not_kwarg_ops = (
+    aten._resize_output_.default,
+    aten.nested_tensor.default,
+    aten.pin_memory.default,
+    aten.is_pinned.default,
+    aten.to.device,
+    aten.to.prim_Device,
+    aten._pin_memory.default,
+)
+
+# Meta tensors give you the ability to run PyTorch code without having to
+# actually do computation through tensors allocated on a `meta` device.
+# Because the device is `meta`, meta tensors do not model device propagation.
+# FakeTensor extends MetaTensors to also carry an additional `fake_device`
+# which tracks devices that would have been used.
+
+
+class FakeTensor(BaseTensor):
+    fake_device: torch.device
+
+    @staticmethod
+    def __new__(cls, elem, device):
+        return super().__new__(cls, elem)
+
+    def __init__(self, elem, device: Union[torch.device, str]):
+        # elem does not need to be recorded, because FakeTensor *is a* elem
+        assert elem.device.type == "meta"
+        device = device if isinstance(device, torch.device) else torch.device(device)
+        assert device.type != "meta"
```

Comment:
You know... it might be ok for the inner device to be meta lol

Reply:
That seems like something has gone wrong (lost what the actual device is)... maybe if it proves to be needed in the future we can change